src/ngraph/pass/algebraic_simplification.cpp

```diff
@@ -55,6 +59,71 @@ static std::shared_ptr<pattern::op::Label>
     return std::dynamic_pointer_cast<pattern::op::Label>(matcher->pattern_node()->get_argument(1));
 }
 
+static bool simplify_concat(std::shared_ptr<Node> n)
+{
+    NGRAPH_DEBUG << "In simplify_concat for " << n->get_name();
+
+    std::shared_ptr<Node> goe;
+
+    auto lgoe = std::make_shared<pattern::op::Label>(element::i32, Shape{2, 1});
+
+    auto slice =
+        std::make_shared<op::Slice>(lgoe, Coordinate{0, 0}, Coordinate{2, 1}, Strides{1, 1});
+
+    auto reshape_pred = [](std::shared_ptr<Node> r) {
+        return std::dynamic_pointer_cast<op::Reshape>(r) != nullptr;
+    };
+
+    auto skip_reshape = std::make_shared<pattern::op::Skip>(slice, reshape_pred);
+
+    auto matcher = std::make_shared<pattern::Matcher>(skip_reshape, nullptr);
+
+    for (auto carg : n->get_arguments())
+    {
+        if (!matcher->match(carg))
+        {
+            NGRAPH_DEBUG << carg->get_name() << " doesn't match";
+            return false;
+        }
+
+        if (goe)
+        {
+            if (goe != matcher->get_pattern_map()[lgoe])
+            {
+                NGRAPH_DEBUG << goe->get_name() << " doesn't match "
+                             << matcher->get_pattern_map()[lgoe]->get_name();
+                return false;
+            }
+        }
+        else
+        {
+            goe = matcher->get_pattern_map()[lgoe];
+            NGRAPH_DEBUG << "setting goe to " << goe->get_name();
+        }
+
+        auto it = carg;
+        while (it != goe)
+        {
+            if (carg->get_users().size() > 1)
+            {
+                NGRAPH_DEBUG << carg->get_name() << " has more than one user";
+                return false;
+            }
+
+            it = it->get_argument(0);
+        }
+    }
+
+    if (!std::dynamic_pointer_cast<op::GetOutputElement>(goe))
+    {
+        NGRAPH_DEBUG << goe->get_name() << " isn't GOE ";
```

Comment:
@diyessi 
This isn't specific to GOE... we should be able to do it for any op but I thought it might be better if we make this opt more targeted?